kasa/klapprotocol.py

```diff
@@ -0,0 +1,500 @@
+        """Query the device retrying for retry_count on failure."""
+        if isinstance(request, dict):
+            request = json_dumps(request)
+            assert isinstance(request, str)
+
+        async with self.query_lock:
+            return await self._query(request, retry_count)
+
+    async def _query(self, request: str, retry_count: int = 3) -> Dict:
+        for retry in range(retry_count + 1):
+            try:
+                return await self._execute_query(request, retry)
+            except aiohttp.ClientConnectionError as ex:
+                raise SmartDeviceException(
+                    f"Unable to connect to the device: {self.host}: {ex}"
+                )
+            except TimeoutError as ex:
+                raise SmartDeviceException(
+                    f"Unable to connect to the device, timed out: {self.host}: {ex}"
+                )
+            except AuthenticationException as auex:
+                _LOGGER.debug("Unable to authenticate with %s, not retrying", self.host)
+                raise auex
+            except Exception as ex:
+                if retry >= retry_count:
+                    _LOGGER.debug("Giving up on %s after %s retries", self.host, retry)
+                    raise SmartDeviceException(
+                        f"Unable to connect to the device: {self.host}: {ex}"
+                    )
+                continue
+
+        # make mypy happy, this should never be reached..
+        raise SmartDeviceException("Query reached somehow to unreachable")
+
+    async def _execute_query(self, request: str, retry_count: int) -> Dict:
+        timeout = aiohttp.ClientTimeout(total=self.timeout)
+
+        async with aiohttp.ClientSession(
+            cookie_jar=self.jar, timeout=timeout
+        ) as session:
+            if not self.handshake_done:
+                try:
+                    await self.perform_handshake(session)
+
+                except AuthenticationException as auex:
+                    _LOGGER.debug(
+                        f"Unable to complete handshake for device {self.host}, authentication failed"
+                    )
+                    raise auex
+
+            # Check for mypy
+            if self.encryption_session is not None:
+                payload, seq = self.encryption_session.encrypt(request.encode())
+
+            url = f"http://{self.host}/app/request"
+
+            cookie = self.jar.filter_cookies(url).get(self.TP_SESSION_COOKIE_NAME)
+
+            response_status, response_data = await self.session_post(
+                session, url, params={"seq": seq}, data=payload
+            )
+
+            if response_status != 200:
+                _LOGGER.error(
+                    f"Query failed after succesful authentication at {datetime.datetime.now()}.  Host is {self.host}, Session cookie is {cookie}, Retry count is {retry_count}, Sequence is {seq}, Response status is {response_status}, Request was {request}"
```

Comment:
https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-not-lazy.html

Reply:
Fixed